### DISCUSSION TITLE
How to prevent the transparent png from casting shadow

### DISCUSSION THREAD

**Discussion by HSDSZ** [bnpr/Malt#337] [Q&A]
The transparent part of a png should not cast any shadow. how should I set to make the shadow a circle?
![image](https://user-images.githubusercontent.com/36077492/175273409-8a2b7284-aa6a-4934-8205-b4f8656549d1.png)

**Comment by pragma37** [comment]
You need to set the opacity in the pre-pass:

![imagen](https://user-images.githubusercontent.com/36610999/175313654-bf96b462-3319-4ec8-83e7-ac82c358591f.png)

Maybe we should have a "simplified" output node that sets the right values for the pre-pass and the main-pass?

**Reply by HSDSZ** [reply]
But cuurently, there are several parameters related with transparent
1 alpha of the mainpass color
2 opacity of the prepass
3 tansparent in the material tab
4 transparent single layer in the material tab

A simplified version might help (but i dont know what is the best solution). Or a comparison of these parameters?

**Reply by pragma37** [reply]
Let's start by the easy ones:

- The *Transparency* option in the material settings enables *alpha blended* transparency. Otherwise only *alpha cut* transparency is enabled.

- *Transparency Single Layer*  means self-overlapping parts of the mesh are not taken into account.

Here, left and right monkeys have the exact same settings except *Transparency Single Layer* is disabled in the left and enabled on the right:
![imagen](https://user-images.githubusercontent.com/36610999/176489344-3422b30d-0ed0-4509-a194-4d4993eafe67.png)

Now the tricky part:

- Opacity in the Pre Pass is used for *alpha cut* transparency (ie. if your opacity is 0, then the surface will be fully transparent, no matter the alpha value of your Main Pass outputs). And it's also used for the Shadow Map transparency.

- Alpha in any Main Pass output is just the alpha value of the output. Since the Color output is connected to the Render Layer output, transparent layers are composited accordingly.
But you can actually set the alpha value for "opaque" objects as well.
Here the monkeys are opaque objects and hide the objects behind them accordingly, but since their alpha is less than 1 the background color can be seen through them since the Background Shader only takes the alpha value into account:
![imagen](https://user-images.githubusercontent.com/36610999/176493300-747a2af8-e9f3-4116-9185-18fba8c5ad8c.png)


The thing is, the Color output in the Main Pass is just a Custom Output, so you can remove it, rename it, have multiple color outputs, or whatever. So the engine can't assume it's there or that it has any specific meaning.

I know this is "weird", but it is the result of providing much more freedom to the user than any other engine. There are probably ways to have "the best of both worlds" but it requires design iteration.
